File: BasePriciples/Liskov_Substitution_Principle/GoVersion/main/RectangleDemo.go
package main

import (
	LSgoVersion "LSGoVersion"
)

func main() {
	r := new(LSgoVersion.Rectangle)
	r.SetLength(20)
	r.SetWidth(10)
	resize(r)
	printLengthAndWidth(r)

	s := new(LSgoVersion.Square)
	s.SetSide(10)
	//resize(s) 无法使用，假如正方形还继承着长方形的话，这就会卡死，现在不会了
	printLengthAndWidth(s)
}

func resize(r *LSgoVersion.Rectangle) {
	for r.GetWidth() <= r.GetLength() {
		r.SetWidth(r.GetWidth() + 1)
	}
}

func printLengthAndWidth(q LSgoVersion.Quadrilateral) {
	println(int(q.GetLength()))
	println(int(q.GetWidth()))
}

File: Java/src/main/java/com/design/pattern/basePriciples/Liskov_Substitution_Principle/RectangleDemo.java
package com.design.pattern.basePriciples.Liskov_Substitution_Principle;

public class RectangleDemo {
    public static void main(String[] args) {
        Rectangle r = new Rectangle();
        r.setLength(20);
        r.setWidth(10);
        resize(r);
        printLengthAndWidth(r);
    }

    public static void resize(Rectangle rectangle) {
        while(rectangle.getWidth() <= rectangle.getLength()) {
            rectangle.setWidth(rectangle.getWidth() + 1);
        }
    }

    public static void printLengthAndWidth(Quadrilateral quadrilateral) {
        System.out.println(quadrilateral.getLength());
        System.out.println(quadrilateral.getWidth());
    }
}
